content/articles/gulp4-primer.md

```diff
@@ -0,0 +1,131 @@
+    at Gulp.task (/home/vincent/Code/ng-boilerplate/node_modules/gulp/node_modules/undertaker/lib/task.js:14:8)
+....
+```
+This comes from the fact that gulp.task changed and the 3 parameters signature was removed.
+In short, tasks that look like the one below will need to be changed to remove the [] parameter:
+```js
+gulp.task('default', ['build'], function () {
+  return runSequence(['watch', 'karma-watch']);
+});
+```
+To give a quick background to people not familiar to gulp, the list parameter is a list of tasks that need to be ran before running that task, in this case it will call the build task before running watch and karma-watch in order (runSequence is another thing that can be changed but we will get back to that later).
+To solve that very common issue of task dependencies, 2 methods were added:
+
+- gulp.series: will run the tasks in order
+- gulp.parallel: will run the tasks in parallel
+
+Let's fix our error by replacing those dependent tasks, for example for the default task:
+
+```js
+gulp.task(
+  'default',
+  gulp.series('build', gulp.parallel('browser-sync', 'watch', 'karma-watch'))
+);
+```
+
+In that case we first want to run build, and then browser-sync, watch and karma-watch in parallel. Easy (the commit below will contain a buggy version since I did put browser-sync in the series by mistake, *easy*)
+Now trying to run the build task, I get another error:
+```js
+[15:09:07] TypeError: undefined is not a function
+    at /home/vincent/Code/ng-boilerplate/node_modules/run-sequence/index.js:18:22
+    at Array.forEach (native)
+    at verifyTaskSets (/home/vincent/Code/ng-boilerplate/node_modules/run-sequence/index.js:12:11)
+```
+Since gulp 4 has ways to run things sequentially, let's use that.
+```js
+// before
+gulp.task('build', function(){
+  return runSequence(
+    'clean',
+    ['sass', 'copy-assets', 'ts-compile', 'templates', 'copy-vendor'],
+    'index'
+  );
+)};
+// after
+gulp.task(
+  'build',
+  gulp.series(
+    'clean',
+    gulp.parallel('sass', 'copy-assets', 'ts-compile', 'templates', 'copy-vendor'),
+    'index'
+  )
+);
+```
+
+Note: I had to modify some tasks like the browser-sync since its API changed and had to make a [patch](https://github.com/dlmanning/gulp-sass/pull/207) to gulp-sass to update the node-sass version.
+You can see the commit [here](https://github.com/Keats/ng-boilerplate/commit/91a05401c6ea532467bc00b5a6e54fd95b6b0eaf).
+Looking at the [build](https://travis-ci.org/Keats/ng-boilerplate/builds/54274589), it still fails because of a missing selenium jar so a protractor error.
+My elegant solution was the [following](https://github.com/Keats/ng-boilerplate/commit/d999db2442ac72fc440e8ff9cccd8348c533c72d), as in removing protractor from travis because I'm not using it at point anyway.
+
+### Small cleanup
+This part is not strictly related to gulp4 but contains some of the things I like to do to have a clean gulpfile.
```

Comment:
Maybe "The following is" instead of "This part". I was puzzling which part you meant for a while.